core/model/src/geometry_sampled_field.cpp

```diff
@@ -172,44 +186,66 @@ static std::vector<QRgb> setImagePixels(
   }
-  SPDLOG_ERROR("Importing sampled field of {} samples of type {}",
+  SPDLOG_DEBUG("Importing sampled field of {} samples of type {}",
                values.size(), common::decltypeStr<T>());
-  if (static_cast<int>(values.size()) != sampledField->getSamplesLength()) {
-    SPDLOG_WARN("Number of samples {} doesn't match SamplesLength {}",
-                values.size(), sampledField->getSamplesLength());
-  }
-  std::size_t iCol = 0;
-  auto iter = colours.begin();
-  for (const auto *sampledVolume : sampledVolumes) {
-    auto matches = getMatchingSampledValues(values, sampledVolume);
-    if (std::find(matches.cbegin(), matches.cend(), true) != matches.cend()) {
-      auto col = common::indexedColours()[iCol].rgb();
-      SPDLOG_DEBUG("  {}/{} -> colour {:x}", sampledVolume->getId(),
-                   sampledVolume->getDomainType(), col);
-      setMatchingPixelsToColour(img, matches, col);
+  if (common::isItIndexes(values, importedSampledFieldColours.size())) {
+    setImportedColoursToImage(img, values, importedSampledFieldColours);
+    auto iter = colours.begin();
+    for (const auto *sampledVolume : sampledVolumes) {
+      int sampledValue = sampledVolume->getSampledValue();
+      auto col = importedSampledFieldColours[sampledValue];
+      SPDLOG_DEBUG("importedSampledFieldColours  {}/{} -> colour {:x}",
+                   sampledVolume->getId(), sampledVolume->getDomainType(), col);
       *iter = col;
-      ++iCol;
+      ++iter;
+    }
+  }
+
+  else {
+    if (static_cast<int>(values.size()) != sampledField->getSamplesLength()) {
+      SPDLOG_WARN("Number of samples {} doesn't match SamplesLength {}",
+                  values.size(), sampledField->getSamplesLength());
+    }
+    std::size_t iCol = 0;
+    auto iter = colours.begin();
+    for (const auto *sampledVolume : sampledVolumes) {
+      auto matches = getMatchingSampledValues(values, sampledVolume);
+      if (std::find(matches.cbegin(), matches.cend(), true) != matches.cend()) {
+        auto col = common::indexedColours()[iCol].rgb();
+        SPDLOG_WARN("Color {} is {}", iCol, col);
+        int sampledValue = sampledVolume->getSampledValue();
+        // if (sampledValue < importedSampledFieldColours.size()) {
+        //   col = importedSampledFieldColours[sampledValue];
+        //   SPDLOG_WARN("  -> Using importedcompartmentColour {}", col);
+        // }
+        SPDLOG_DEBUG("  {}/{} -> colour {:x}", sampledVolume->getId(),
+                     sampledVolume->getDomainType(), col);
+        setMatchingPixelsToColour(img, matches, col);
+        *iter = col;
+        ++iCol;
+      }
+      ++iter;
     }
-    ++iter;
   }
   return colours;
 }
 
 static std::vector<std::pair<std::string, QRgb>> getCompartmentIdAndColours(
     const std::vector<const libsbml::Compartment *> &compartments,
-    const std::vector<QRgb> &compartmentColours) {
+    const std::vector<QRgb> &sampledFieldColours) {
```

Comment:
I would keep the original name to avoid confusion - at this point we are really talking about the colour of each compartment
```suggestion
    const std::vector<QRgb> &compartmentColours) {
```